esp/esdllib/README.md

```diff
@@ -0,0 +1,1608 @@
+```
+
+In some cases you may see XML or xpath parsing errors when using the single quotes depending on your workflow (_establish precisely when this occurs_). This may be caused by unescaped apostrophes, which you can remedy like this:
+
+```xml
+<es:set-value target="name" value="&apos;foobar&apos;"/>
+```
+
+So far we've concluded that escaping is not required in these attributes:
+
+* `getDataSection()` function inside entry point `@source` attribute
+* `set-value[@value]` attribute
+
+### Implicit Variables
+
+These special-case variables are implicitly defined by the scripting engine. They give scripts access to characteristics of the current transaction and context. For example, say we've defined an ESP service named `WsPersonInfo` with a method named `GetPersonInfo` that has an interface defined as:
+
+```xml
+    <GetPersonInfoRequest>
+        <FirstName/>
+        <LastName/>
+    </GetPersonInfoRequest>
+```
+
+And is bound to a back end with this configuration:
+
+```xml
+    <Method name="GetPersonInfo"  url="https://backend.roxie.com:1234" mode="ESDL" queryname="PersonInfo_Services.GetPersonInfo" status="available"/>
+```
+
+* `$method`
+: The name of the ESP method invoked by the current transaction request. In our example this is the string *GetPersonInfo*.
+
+* `$request`
+: The name of the ESP request structure at the root of the current transaction request. In our example this is the string *GetPersonInfoRequest*.
+
+* `$service`
+: The name of the service containing the current transaction's method. In our example this is the string *WsPersonInfo*.
+
+* `$query`
+: The name of the back end roxie query. In our example this is the string *PersonInfo_Services.GetPersonInfo*.
+
+### System Parameters
+
+Each of these parameters gives the script access to some information about the current execution context. In order to reference a parameter, it must be declared in the script like so:
+
+```xml
+<es:param name="clientversion" select="'unknown'"/>
+```
+
+If the parameter is undefined, then it is assigned the value of the `@select` XPath expression.
+
+* `$clientversion` : The client interface version specified in the incoming request.
+* `$espUserName` : The username of the user authenticated to run the current transaction.
+* `$espUserPeer` : The IP address of host originating the current transaction.
+* `$espUserRealm` : The 'Company' or parent entity to the currently authenticated user.
+* `$espUserStatus` : The status of the currently authenticated user (eg *Active*, *Suspended*, etc., as defined by the security plugin and back end in use)
+
+In addition to the standard parameters above, any *Secure User Settings* configured as part of your service deployment are made available as parameters for you to declare and use in scripts. In configmgr, these settings are under ESP Process|ESP Service Bindings|Secure User Setting section. When configuring as an esdl application with a yaml file, it is in the authNZ.auth_map section
```

Comment:
But how do we document what currently works?  And in a way that will work even if we extend it later?

Reply:
Right now, secure user settings cannot be documented as available to scripting in an open source document, because only proprietary managers implement non-standard behavior to make them available. No platform defined manager is even aware of these settings.

The settings either shouldn't be mentioned yet, or we make the relatively quick change to iterate the settings resource list and define the values so the the current segment is correct. The idea of a user section could be viewed as a future upgrade that would just provide a second way to access the same values but could also include more user data that is neither property nor setting.

Reply:
Updated this section to only refer to Secure User Properties, and opened a ticket to investigate & possibly implement the other ideas mentioned: https://track.hpccsystems.com/browse/HPCC-27881